liblll/CodeFragment.cpp

```diff
@@ -523,14 +523,24 @@ void CodeFragment::constructOperation(sp::utree const& _t, CompilerState& _s)
 			requireSize(1);
 			requireDeposit(0, 1);
 
+			/// Save starting offset on stack.
 			m_asm.append(Instruction::MSIZE);
-			m_asm.append(u256(0));
-			m_asm.append(u256(1));
+
+			/// Calculate needed memory size.
+			m_asm.append(u256(31));
 			m_asm.append(code[0].m_asm, 1);
-			m_asm.append(Instruction::MSIZE);
+			m_asm.append(Instruction::DUP3);
 			m_asm.append(Instruction::ADD);
+
+			/// Trigger memory expansion
 			m_asm.append(Instruction::SUB);
-			m_asm.append(Instruction::MSTORE8);
+			auto end = m_asm.newTag();
+			m_asm.append(u256(0));
+			m_asm.append(Instruction::DUP2);
+			m_asm.append(Instruction::LT);
+			m_asm.appendJumpI(end);
+			m_asm.append(Instruction::MLOAD);
+			m_asm.append(end);
 
```

Comment:
We need to add a POP here after the end as both branches leave something on the stack unnecessarily (the output of the MLOAD or the result of the subtraction).

Even with these mods, the code allocs too much for (alloc 32) and not enough for N<31.  I'll have a go a reworking it.